feat: adding google malware detection

jsdevel opened the pull request:
- [x] Test ipfs malware sample from google with lookup and evaluate.

vasco-santos commented on packages/edge-gateway/src/gateway.js:
```diff
@@ -110,6 +110,13 @@ export async function gatewayGet (request, env, ctx) {
     }
+
+    if (winnerGwResponse.headers.get('content-type') === 'text/html') {
+      // we don't wait for google because it adds to the overall latency, but in the
+      // background we'll update KV so future requests coming in will properly be
+      // blocked as CIDs/malware are flagged.
+      fetchGoogleMalwareDetection(winnerUrl, cid, env)
+    }
```
We need to add a safeguard here to prevent recursive requests (more info on https://hackmd.io/@vasco-santos/rJWZsz-Cc point 1 flow notes)

vasco-santos commented on packages/edge-gateway/src/gateway.js:
```diff
@@ -110,6 +110,13 @@ export async function gatewayGet (request, env, ctx) {
       return new Response(reason || '', { status: status || 410 })
     }
+
+    if (winnerGwResponse.headers.get('content-type') === 'text/html') {
```
Let's use `includes` given content type can be values such as `text/html; charset=utf-8`

vasco-santos commented on packages/edge-gateway/src/gateway.js:
```diff
@@ -274,6 +281,37 @@ async function getFromDenyList (datastore, cid) {
+/**
+ * Fetch malware state from google's malware detection API and update our DENYLIST.
+ *
+ * TODO: notify badbits
```
We won't be able to notify badbits from here (at least in the current scenario where it requires email to be sent).  We should just rely on logs

vasco-santos reviewed (changes requested):
Thanks for the first iteration @jsdevel 

Let's please create a package from the start as previously talked. Once we hook up tests, KV to store results / KV Locker it will be a bunch of logic that should be isolated. Also, we will want the logs for metrics + trigger badbits mail

For the package, we want a worker quite similar to this one with:
- wrangler.toml
- index.js
- cors.js
- error-handler.js
- env.js

vasco-santos commented on packages/cid-verifier/scripts/cli.js:
```diff
@@ -0,0 +1,21 @@
+  .describe('Build the worker.')
+  .option('--env', 'Environment', env)
+  .action(buildCmd)
+  .command('heartbeat', 'Ping opsgenie heartbeat')
+  .option('--token', 'Opsgenie Token')
+  .option('--name', 'Heartbeat Name')
+  .action(heartbeatCmd)
```
```suggestion
```

vasco-santos commented on packages/cid-verifier/scripts/cli.js:
```diff
@@ -0,0 +1,21 @@
+import sade from 'sade'
+
+import { buildCmd } from './build.js'
+import { heartbeatCmd } from './heartbeat.js'
```
```suggestion
```

we don't need hearbeat. It is used for tracking if cron jobs are running and creating alerts to opsgeanie

vasco-santos commented on packages/cid-verifier/scripts/heartbeat.js:
```diff
@@ -0,0 +1,14 @@
+import { fetch } from '@web-std/fetch'
+
+export async function heartbeatCmd (opts) {
+  try {
+    await fetch(`https://api.opsgenie.com/v2/heartbeats/${opts.name}/ping`, {
+      headers: {
+        Authorization: `GenieKey ${opts.token}`
+      }
+    })
+  } catch (err) {
+    console.error(err)
+    process.exit(1)
+  }
+}
```
```suggestion
```

vasco-santos commented on packages/cid-verifier/package.json:
```diff
@@ -0,0 +1,42 @@
+    "test": "npm run test:setup && npm-run-all -p -r mock:ipfs.io mock:cf-ipfs.com test:worker",
+    "test:worker": "ava --verbose test/*.spec.js",
+    "test:setup": "npm run build",
+    "mock:ipfs.io": "smoke -p 9082 test/mocks/ipfs.io",
+    "mock:cf-ipfs.com": "smoke -p 9083 test/mocks/cf-ipfs.com"
```
I know you did not get to this yet, but since I am here this should be `mock:google-api`

jsdevel replied:
yea, i'm still in active development on it

vasco-santos commented on packages/cid-verifier/src/bindings.d.ts:
```diff
@@ -0,0 +1,57 @@
+
+export {}
+
+// CF Analytics Engine types not available yet
+export interface AnalyticsEngine {
+  writeDataPoint(event: AnalyticsEngineEvent): void
+}
+
+export interface AnalyticsEngineEvent {
+  readonly doubles?: number[]
+  readonly blobs?: (ArrayBuffer | string | null)[]
+}
```
```suggestion
```

not needed, at least for now

vasco-santos commented on packages/cid-verifier/src/bindings.d.ts:
```diff
@@ -0,0 +1,57 @@
+  readonly blobs?: (ArrayBuffer | string | null)[]
+}
+
+export interface EnvInput {
```
a bunch of things here are not needed

vasco-santos commented on packages/cid-verifier/src/env.js:
```diff
@@ -0,0 +1,79 @@
+ * @param {Ctx} ctx
+ */
+export function envAll (request, env, ctx) {
+  env.REQUEST_TIMEOUT = env.REQUEST_TIMEOUT || 20000
+  env.IPFS_GATEWAY_HOSTNAME = env.GATEWAY_HOSTNAME
+  env.IPNS_GATEWAY_HOSTNAME = env.GATEWAY_HOSTNAME.replace('ipfs', 'ipns')
```
```suggestion
```

vasco-santos commented on packages/cid-verifier/src/env.js:
```diff
@@ -0,0 +1,79 @@
+  env.SENTRY_RELEASE = SENTRY_RELEASE
+
+  env.sentry = getSentry(request, env, ctx)
+  env.ipfsGateways = JSON.parse(env.IPFS_GATEWAYS)
+  env.startTime = Date.now()
```
```suggestion
```

vasco-santos reviewed (commented):
Thanks for the new iteration Joe. Added some initial feedback even though I know this is still WIP 😉

vasco-santos commented on packages/cid-verifier/src/bindings.d.ts:
```diff
@@ -0,0 +1,58 @@
+  LOKI_URL?: string;
+  LOKI_TOKEN?: string;
+  DENYLIST: KVNamespace;
+  GOOGLE_MALWARE_RESULTS: KVNamespace;
```
Wondering if we should call this: `CID_VERIFIER`, so that we are better prepared if we had other tools. What do you think?

We did not discuss this yet, but looking at the code remembered we should consider this.

Let's consider that `sourceName` is the source where we get a result (in this case a google API code, the URL would be the nicest, but probably makes the KV key larger than the maximum allowed?). There are two options for how to setup `{key,value}`:
- key = `cid`, value = `[{ source: string, value: any }]`
- key = `cid/${sourceName}`, value  = sourceValue

The first option might be problematic for multiple atomic writes, given [KV is eventually consistent](https://developers.cloudflare.com/workers/learning/how-kv-works/).

The second option does not have write issues, but makes reads harder. We can [list by prefix](https://developers.cloudflare.com/workers/runtime-apis/kv/#listing-keys) (in this case CID) to get all the keys. There is an extra trick we could do, just have the response from the source being both KV value and metadata. If you see kv.list API, they return key + metadata, so we don't need list + n get operations to get what we care about. `metadata` can only have "The serialized JSON representation of the metadata object must be no more than 1024 bytes in length." This should be enough right? (we should document this)

Let me know what you think here @jsdevel  :)

jsdevel replied:
i kinda like a seperate KV named GOOGLE_MALWARE_RESULTS.  i went with cid for the key because the URL isn't really important once we know the cid is bad.  i think 1024 bytes will be enough in the value

vasco-santos replied:
What would be your idea for when we add other services that are not google?

We still need to tackle the KV eventually consistent problem too. Any thoughts on that?

vasco-santos commented on packages/cid-verifier/src/bindings.d.ts:
```diff
@@ -0,0 +1,58 @@
+  BRANCH: string;
+  COMMITHASH: string;
+  SENTRY_RELEASE: string;
+  REQUEST_TIMEOUT: number;
+  IPFS_GATEWAY_HOSTNAME: string;
+  IPNS_GATEWAY_HOSTNAME: string;
+  ipfsGateways: Array<string>;
```
```suggestion
```

vasco-santos commented on packages/cid-verifier/src/bindings.d.ts:
```diff
@@ -0,0 +1,58 @@
+  ipfsGateways: Array<string>;
+  sentry?: Toucan;
+  log: Logging;
+  startTime: number;
```
```suggestion
```

vasco-santos commented on packages/cid-verifier/src/env.js:
```diff
@@ -0,0 +1,76 @@
+ * @param {Ctx} ctx
+ */
+export function envAll (request, env, ctx) {
+  env.REQUEST_TIMEOUT = env.REQUEST_TIMEOUT || 20000
+
```
```suggestion
```

vasco-santos commented on packages/cid-verifier/src/env.js:
```diff
@@ -0,0 +1,76 @@
+  env.SENTRY_RELEASE = SENTRY_RELEASE
+
+  env.sentry = getSentry(request, env, ctx)
+  env.startTime = Date.now()
```
```suggestion
```

vasco-santos commented on packages/cid-verifier/package.json:
```diff
@@ -0,0 +1,50 @@
+    "test": "npm run test:setup && npm-run-all -p -r mock:google-cloud.io test:worker",
+    "test:worker": "ava --verbose test/*.spec.js",
+    "test:setup": "npm run build",
+    "mock:google-cloud.io": "smoke -p 9111 test/mocks/google-cloud.io --set foo"
```
```suggestion
    "mock:google-cloud.io": "smoke -p 9111 test/mocks/google-cloud.io"
```

we don't need this right?

vasco-santos commented on packages/edge-gateway/wrangler.toml:
```diff
@@ -108,8 +108,12 @@ workers_dev = true
 kv_namespaces = [{ binding = "DENYLIST" }]
 
 [env.test.vars]
+# TODO: setup smoke server for tests
+GOOGLE_CLOUD_API_URL = 'https://webrisk.googleapis.com'
+GOOGLE_CLOUD_API_KEY = 'TODO'
```
```suggestion
```

vasco-santos commented on packages/edge-gateway/wrangler.toml:
```diff
@@ -108,8 +108,12 @@ workers_dev = true
 IPFS_GATEWAYS = "[\"http://127.0.0.1:9081\", \"http://localhost:9082\", \"http://localhost:9083\"]"
 GATEWAY_HOSTNAME = 'ipfs.localhost:8787'
 EDGE_GATEWAY_API_URL = 'http://localhost:8787'
+CID_VERIFIER_URL = 'http://127.0.0.1:9111'
```
This should be worker binding like we setup in https://github.com/web3-storage/reads/blob/edge-gateway-v1.0.4/packages/edge-gateway/wrangler.toml#L42-L46

vasco-santos commented on packages/edge-gateway/src/utils/deny-list.js:
```diff
@@ -1,13 +0,0 @@
-import * as uint8arrays from 'uint8arrays'
```
All good here, but don't forget we need to update file `.github/workflows/cron-denylist.yml` for new path. Also docs https://github.com/web3-storage/reads/tree/edge-gateway-v1.0.4/packages/edge-gateway#deny-list should be moved to new package readme

vasco-santos commented on packages/edge-gateway/package.json:
```diff
@@ -11,11 +11,12 @@
     "dev": "miniflare dist/worker.mjs --watch --debug -m",
     "postdev": "node scripts/cli.js ipfs --stop",
-    "test": "npm run test:setup && npm-run-all -p -r mock:ipfs.io mock:cf-ipfs.com test:worker",
+    "test": "npm run test:setup && npm-run-all -p -r mock:ipfs.io mock:cf-ipfs.com mock:cid-verifier.io test:worker",
```
We shouldn't mock cid verifier like this here. If we use the worker binding (see my comment on wrangler.toml file), we should test using the worker binding miniflare API. For this you can look on how we test the gateway API binding:
- https://github.com/web3-storage/reads/blob/main/packages/edge-gateway/test/utils/miniflare.js#L24-L32
- https://github.com/web3-storage/reads/blob/main/packages/edge-gateway/test/utils/scripts/api.js

This is essentially the way Miniflare enables us to test Worker to Worker interactions. So, in this case, most tests should be in `cid-verifier` package. In edge-gateway, probably just the basic interactions with the external API

vasco-santos reviewed (changes requested):
Thanks for the new iteration. Glad things are shaping up. A couple of general comments I could not have a place to add in review:
- github actions workflow for cid verifier package (can be based on edge-gateway
- CI is having a bad time, looks like pnpm lockfile should be commited? perhaps run on root dir pnpm clean && pnpm i

Also, this is first set of comments and I did not get into the actual code for CID verification. Just all the boilerplate around

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,130 @@
+    // cid not processed so we process it
+
+    // we set the status to pending to limit the number of times we call google
+    env.GOOGLE_MALWARE_RESULTS.put(cid, '"PENDING"')
```
```suggestion
    env.GOOGLE_MALWARE_RESULTS.put(cid, '"PENDING"')
```

I am concerned about KV nature being eventually consistent here. For example, if two concurrent requests happen (one in Europe and one in the USA) there is a possibility of:
1. req1.get()
2. req2.get()
3. req1.put('PENDING')
4. req1.put(result)
5. req2.put('PENDING')
6. req2.put(result)

Eventually it will be fine in the end. But building on my other comment https://github.com/web3-storage/reads/pull/36#discussion_r960765098 `packages/cid-verifier/src/bindings.d.ts`. Probably, we can go even a step further and write a key lock: `cid/google-evaluate-api.lock` with PENDING value

Once we have final value, we write to key `cid/google-evaluate-api` and delete lock. This way, we never have issues with concurrent writes in same key.

Of course this means above we should use KV.list instead of KV.get.

Let me know your thoughts here, perhaps you have other ideas in mind

jsdevel replied:
i kinda feel eventual consistency is safer and potentially serves up more concurrent requests compared to locking.

vasco-santos replied:
I don't agree with safer. What if response from Google fails? This is also an issue with locks though, if it is not cleanned up, but with locks we can safely clean up if it fails, compared to clean up a pending that might delete a concurrent put result.

We don't need concurrent requests, we just want one per CID.

jsdevel replied:
i realized by having DENYLIST we avoid the need for a seperate lock entry.

vasco-santos replied:
The concurrent issue won't be in GET route (where we changed to only check DENYLIST). It will happen for GET (List after refactor we talked), PUT, DELETE of same key in concurrent POST actions

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,130 @@
+        env.GOOGLE_MALWARE_RESULTS.put(cid, await googleCloudResponse.text())
+        return new Response('cid malware detection processed', { status: 203 })
+      } catch (e) {
+        // TODO: best way to handle errors?
```
use `env.log` to send and error to log for now.

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,130 @@
+
+    if (googleCloudResponse.status === 200) {
+      try {
+        env.GOOGLE_MALWARE_RESULTS.put(cid, await googleCloudResponse.text())
```
Consider adding third parameter with `{metadata: ...}` as previously mentioned.

so here in POST, if this is considered malware I think it is easier to just do a new PUT to env.DENYLIST.put

This way, GET only needs to check the denylist because it only cares about if it is bad or not. We internally want to have more information we can share, and that's why it is also useful to have the extra KV.

So, in Denylist we can simply put value { status, reason } like we have in scripts/denylist.js

What do you think?

jsdevel replied:
i kinda think it's safer to seperate google and badbits into seperate KV stores.  for any given cid we know exactly what is stored in the respective KVs.

vasco-santos replied:
Do you mean that you agree with the suggestion, or do you have something else in mind?

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,130 @@
+      }
+    }
+
+    const googleMalwareResults = await getGoogleMalwareResult(cid, env)
+    if (!googleMalwareResults) {
+      return new Response('Verification Not Known', { status: 404 })
+    }
+
+    if (googleMalwareResults === '"PENDING"') {
+      return new Response('Verification Pending', { status: 202 })
+    }
+
+    const parsedGoogleMalwareResults = JSON.parse(googleMalwareResults)
+    if (!Object.keys(parsedGoogleMalwareResults).length) {
+      return new Response('Malware Not Detected', { status: 200 })
+    } else {
+      return new Response('MALWARE DETECTED', { status: 403 })
+    }
+  }
```
We should simply return 200 here based on my comments for POST. Thoughts?

jsdevel commented:
> Thanks for the new iteration. Glad things are shaping up. A couple of general comments I could not have a place to add in review:
> 
> * github actions workflow for cid verifier package (can be based on edge-gateway
> * CI is having a bad time, looks like pnpm lockfile should be commited? perhaps run on root dir pnpm clean && pnpm i
> 
> Also, this is first set of comments and I did not get into the actual code for CID verification. Just all the boilerplate around

i saw that error.  i checked in the pnpm lock file.  i think locally i'm using a later version of pnpm and that's causing it.

vasco-santos commented:
> i saw that error. i checked in the pnpm lock file. i think locally i'm using a later version of pnpm and that's causing it.

let me know if I can help with something there then

vasco-santos commented on packages/cid-verifier/package.json:
```diff
@@ -0,0 +1,47 @@
+  "description": "Verify that a CID's content is safe for the web",
+  "private": true,
+  "type": "module",
+  "main": "./dist/worker.mjs",
```
See this https://github.com/web3-storage/reads/pull/47 . We recently discovered this was a problem, and I fixed for edge-gateway. We need the changes of the PR propagated too here

jsdevel commented:
> The concurrent issue won't be in GET route (where we changed to only check DENYLIST). It will happen for GET (List after refactor we talked), PUT, DELETE of same key in concurrent POST actions

@vasco-santos i updated it so it only attempts to fetch google results if there is currently nothing in GOOGLE_EVALUATE_RESULTS for the given cid.  should be good now.

vasco-santos commented on packages/cid-verifier/package.json:
```diff
@@ -0,0 +1,47 @@
+  "scripts": {
+    "lint": "standard",
+    "build": "tsc && node scripts/cli.js build",
+    "dev": "miniflare dist/worker.mjs --watch --debug -m",
```
```suggestion
    "dev": "miniflare dist/worker.js --watch --debug -m",
```

vasco-santos commented on packages/cid-verifier/scripts/build.js:
```diff
@@ -0,0 +1,60 @@
+    })
+    await cli.releases.uploadSourceMaps(sentryRelease, {
+      include: [path.join(__dirname, '..', 'dist')],
+      ext: ['map', 'mjs']
```
```suggestion
      ext: ['map', 'js']
```

vasco-santos commented on packages/cid-verifier/scripts/build.js:
```diff
@@ -0,0 +1,60 @@
+    entryPoints: [path.join(__dirname, '..', 'src', 'index.js')],
+    bundle: true,
+    format: 'esm',
+    outfile: path.join(__dirname, '..', 'dist', 'worker.mjs'),
```
```suggestion
    outfile: path.join(__dirname, '..', 'dist', 'worker.js'),
```

vasco-santos commented on packages/cid-verifier/src/bindings.d.ts:
```diff
@@ -0,0 +1,58 @@
+
+export {};
+
+// CF Analytics Engine types not available yet
+export interface AnalyticsEngine {
+  writeDataPoint(event: AnalyticsEngineEvent): void;
+}
+
+export interface AnalyticsEngineEvent {
+  readonly doubles?: number[];
+  readonly blobs?: (ArrayBuffer | string | null)[];
+}
```
This needs to be removed

vasco-santos commented on packages/cid-verifier/src/env.js:
```diff
@@ -0,0 +1,73 @@
+    debug: false,
+    environment: env.ENV || 'dev',
+    rewriteFrames: {
+      // strip . from start of the filename ./worker.mjs as set by cloudflare, to make absolute path `/worker.mjs`
+      iteratee: (frame) => ({
+        ...frame,
+        // @ts-ignore
+        filename: frame.filename.substring(1)
+      })
```
```suggestion
      // sourcemaps only work if stack filepath are absolute like `/worker.js`
      root: '/'
```

vasco-santos commented on packages/cid-verifier/src/bindings.d.ts:
```diff
@@ -0,0 +1,55 @@
+export interface EnvInput {
+  ENV: string;
+  DEBUG: string;
+  GOOGLE_EVALUATE_SAFE_CONFIDENCE_LEVELS: Array<string>;
```
Does this get in as a binding?

vasco-santos replied:
looking after everything, it looks like we just not need this?

vasco-santos replied:
Let;s sync on this than https://cloud.google.com/web-risk/docs/reference/rest/v1eap1/TopLevel/evaluateUri#confidencelevel

vasco-santos commented on packages/cid-verifier/src/index.js:
```diff
@@ -0,0 +1,52 @@
+router
+  .all('*', envAll)
+  .get('/version', withCorsHeaders(versionGet))
+  .get('/verification', withCorsHeaders(verificationGet))
+  .post('/verification', withCorsHeaders(verificationPost))
```
What do you think on having this as just `/`?

Considering we will deploy this as `cid-verifier.dag.haus`, API endpoint with `/verification` seems a bit weird to me. We will (in the future) want a way to read from KV info too. So thinking about:

```
GET /
POST /
GET /metadata (future)
```

(this is future work, not here, just to help with semantics) Metadata would return all the information we know from all the services we already requested info

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,137 @@
+    const denyListResource = await getFromDenyList(cid, env)
+    if (denyListResource) {
+      const { status } = JSON.parse(denyListResource)
+      if (status === 451) {
+        return new Response('BLOCKED FOR LEGAL REASONS', { status: 451 })
+      } else {
+        return new Response('MALWARE DETECTED', { status: 403 })
+      }
```
```suggestion
      if (status === 451) {
        return new Response('BLOCKED FOR LEGAL REASONS', { status })
      } else {
        return new Response('MALWARE DETECTED', { status })
      }
```

Probably just keeps status as is?

vasco-santos commented on packages/cid-verifier/test/utils/miniflare.js:
```diff
@@ -0,0 +1,36 @@
+
+  return new Miniflare({
+    envPath,
+    scriptPath: 'dist/worker.mjs',
```
```suggestion
    scriptPath: 'dist/worker.js',
```

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,137 @@
+    const googleMalwareResult = await getGoogleMalwareResult(cid, env)
+
+    if (!googleMalwareResult) {
+      env.GOOGLE_EVALUATE_RESULTS.put(cid, 'PENDING')
```
Let's put everything within the same scope. So, move the put inside `fetchGoogleMalwareResults`. Also:
- this should have await
- this should have key as previously discussed: `cid/${google-service-id-name}`   https://hackmd.io/@vasco-santos/rJWZsz-Cc
- (thought we would go with the lock?)

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,137 @@
+    throw new Error('db not ready')
+  }
+
+  return await pRetry(
+    () => datastore.get(cid),
+    { retries: 5 }
+  )
```
This should do a KV list operation with cid prefix to see if we have any result: either .lock , or a `cid/service` kind of key with results already available

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,137 @@
+  if (googleCloudResponse.status === 200) {
+    try {
+      const evaluateJson = await googleCloudResponse.json()
+      env.GOOGLE_EVALUATE_RESULTS.put(cid, JSON.stringify(evaluateJson))
```
await, correct key, add to metadata for easy listing in the future

vasco-santos commented on packages/cid-verifier/src/bindings.d.ts:
```diff
@@ -0,0 +1,55 @@
+  LOKI_URL?: string;
+  LOKI_TOKEN?: string;
+  DENYLIST: KVNamespace;
+  GOOGLE_EVALUATE_RESULTS: KVNamespace;
```
```suggestion
  CID_VERIFIER_RESULTS: KVNamespace;
```

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,137 @@
+
+  if (!resultsStored) {
+    // delete PENDING so we can try again
+    env.GOOGLE_EVALUATE_RESULTS.delete(cid)
```
await, lock key

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,137 @@
+      const evaluateJson = await googleCloudResponse.json()
+      env.GOOGLE_EVALUATE_RESULTS.put(cid, JSON.stringify(evaluateJson))
+      // @ts-ignore
+      if (evaluateJson?.scores?.some(score => !env.GOOGLE_EVALUATE_SAFE_CONFIDENCE_LEVELS.includes(score.confidenceLevel))) {
```
Can we document this line logic for better readability?

vasco-santos commented on packages/cid-verifier/test/utils/miniflare.js:
```diff
@@ -0,0 +1,56 @@
+    buildCommand: undefined,
+    wranglerConfigEnv: 'test',
+    modules: true,
+    mounts: {
```
This still needs to go away, mounts and service bindings, given cid-verifier does not have any dependency from other service

vasco-santos commented on packages/cid-verifier/test/utils/scripts/api.js:
```diff
@@ -0,0 +1,24 @@
+/* eslint-env serviceworker, browser */
+
+export default {
+  /**
+   * @param {Request} request
+   */
+  async fetch (request) {
+    // TODO: hardcoded until miniflare supports Worker bindings
+    if (
+      request.url
+        .toString()
+        .includes('bafybeic2hr75ukgwhnasdl3sucxyfedfyp3dijq3oakzx6o24urcs4eige')
+    ) {
+      return new Response('Hello perma cache!', {
+        status: 200,
+        headers: { 'Content-Type': 'text/plain' }
+      })
+    }
+
+    return new Response('Not Found', {
+      status: 404
+    })
+  }
+}
```
this still needs to go away

vasco-santos commented on packages/edge-gateway/wrangler.toml:
```diff
@@ -105,11 +103,17 @@ name = "PUBLIC_RACE_STATUS_CODE"
 workers_dev = true
-kv_namespaces = [{ binding = "DENYLIST" }]
+
+[[env.production.services]]
```
We need staging and production. And please put it in order with the other configs for each environment

vasco-santos commented on packages/cid-verifier/wrangler.toml:
```diff
@@ -0,0 +1,55 @@
+
+# PROD!
+[env.production]
+# name = "dotstorage-edge-gateway-production"
```
```suggestion
# name = "dotstorage-cid-verifier-production"
```

vasco-santos commented on packages/cid-verifier/wrangler.toml:
```diff
@@ -0,0 +1,55 @@
+# name = "gateway-nft-storage-staging"
+account_id = "fffa4b4363a7e5250af8357087263b3a" # Protocol Labs CF account
+routes = [
+  "cid-verifier.dag.haus/*"
```
```suggestion
  "cid-verifier-staging.dag.haus/*"
```

vasco-santos commented on packages/cid-verifier/wrangler.toml:
```diff
@@ -0,0 +1,55 @@
+
+# Staging!
+[env.staging]
+# name = "gateway-nft-storage-staging"
```
```suggestion
# name = "dotstorage-cid-verifier-staging"
```

vasco-santos commented on packages/edge-gateway/wrangler.toml:
```diff
@@ -105,11 +103,17 @@ name = "PUBLIC_RACE_STATUS_CODE"
+[[env.production.services]]
+binding = "CID_VERIFIER"
+type = "service"
+service = "cid-verifier-api"
```
```suggestion
service = "dotstorage-cid-verifier-production"
```

and staging on the other one

vasco-santos commented on packages/edge-gateway/wrangler.toml:
```diff
@@ -105,11 +103,17 @@ name = "PUBLIC_RACE_STATUS_CODE"
 [env.test.vars]
 IPFS_GATEWAYS = "[\"http://127.0.0.1:9081\", \"http://localhost:9082\", \"http://localhost:9083\"]"
 GATEWAY_HOSTNAME = 'ipfs.localhost:8787'
+CID_VERIFIER_URL = 'http://cid-verifier.localhost:8787'
```
We need this in prod and staging envs too

vasco-santos commented on packages/edge-gateway/test/malware.spec.js:
```diff
@@ -0,0 +1,44 @@
+  await t.context.container?.stop()
+})
+
+test('Blocks access to a CID that is malicious', async (t) => {
```
Would still love this test for optimal coverage

vasco-santos commented on packages/edge-gateway/src/gateway.js:
```diff
@@ -3,13 +3,13 @@
 import pAny, { AggregateError } from 'p-any'
 import pDefer from 'p-defer'
-import pRetry from 'p-retry'
+// import pRetry from 'p-retry'
```
```suggestion
```

vasco-santos commented on packages/edge-gateway/src/gateway.js:
```diff
@@ -3,13 +3,13 @@
 import { getCidFromSubdomainUrl } from './utils/cid.js'
 import { getHeaders } from './utils/headers.js'
-import { toDenyListAnchor } from './utils/deny-list.js'
+// import { toDenyListAnchor } from './utils/deny-list.js'
```
```suggestion
```

vasco-santos commented on packages/edge-gateway/src/gateway.js:
```diff
@@ -105,10 +103,16 @@ export async function gatewayGet (request, env, ctx) {
-      const { status, reason } = JSON.parse(denyListResource)
-      return new Response(reason || '', { status: status || 410 })
+
+    const cidVerificationResponse = await env.CID_VERIFIER.fetch(`${env.CID_VERIFIER_URL}/verification?cid=${resourceCid}`)
```
Please rename this to `cidResourceVerificationResponse` for better readability that this is resource cid from etag

vasco-santos commented on packages/edge-gateway/test/malware.spec.js:
```diff
@@ -0,0 +1,44 @@
+  const res = await mf.dispatchFetch(`https://${cid}.ipfs.localhost:8787`)
+  t.is(res.status, 403)
+  t.is(await res.text(), 'malicious')
+})
```
We also need to make sure a test happens for calling `env.CID_VERIFIER.fetch(`${env.CID_VERIFIER_URL}/verification?cid=${resourceCid}&url=${encodeURIComponent(request.url)}`, { method: 'POST'`

vasco-santos commented on packages/cid-verifier/wrangler.toml:
```diff
@@ -0,0 +1,55 @@
+routes = [
+  "cid-verifier.dag.haus/*"
+]
+kv_namespaces = [{ binding = "DENYLIST", id = "f4eb0eca32e14e28b643604a82e00cb3" }]
```
created staging KV:

```
🌀 Creating namespace with title "dotstorage-cid-verifier-staging-staging-CID_VERIFIER_RESULTS"
✨ Success!
Add the following to your configuration file in your kv_namespaces array under [env.staging]:
{ binding = "CID_VERIFIER_RESULTS", id = "c742e87b92da4a77bdce7de8105778ff" }
```

vasco-santos commented on packages/cid-verifier/wrangler.toml:
```diff
@@ -0,0 +1,72 @@
+  "*.ipns.dag.haus/*",
+  "*.ipns.dag.haus"
+]
+kv_namespaces = [{ binding = "DENYLIST", id = "785cf627e913468ca5319523ae929def" }]
```
Created prod KV:

```
🌀 Creating namespace with title "dotstorage-cid-verifier-production-production-CID_VERIFIER_RESULTS"
Retrieving cached values for userId from node_modules/.cache/wrangler
✨ Success!
Add the following to your configuration file in your kv_namespaces array under [env.production]:
{ binding = "CID_VERIFIER_RESULTS", id = "9be91fbb7fe34afea6629732981fc8e9" }
```

vasco-santos commented on packages/cid-verifier/wrangler.toml:
```diff
@@ -0,0 +1,55 @@
+workers_dev = true
+kv_namespaces = [
+  { binding = "DENYLIST" },
+  { binding = "GOOGLE_EVALUATE_RESULTS" }
```
```suggestion
  { binding = "CID_VERIFIER_RESULTS" }
```

jsdevel commented:
> New set of changes. Please also add README file, gather a copy from edge-gateway and basically need specific secrets for here specified there

i'll add a README once the approach is finalized

jsdevel commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,143 @@
+  }
+
+  return await pRetry(
+    () => datastore.get(key),
```
i tried using `.list()` but it wasn't returning metadata.  tried adding metadata in the `.put()` calls but no dice.  have we used `.list()` successfully in other repos?

jsdevel replied:
figured it out.  had to update the tests.  working now!

jsdevel commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,146 @@
+    const cidVerifyResults = await getResults(cid, env)
+    // @ts-ignore
+    const googleEvaluateResult = cidVerifyResults[resultKey]
+    // @ts-ignore
```
it would be nice to refactor edge-gateway and cid-verifier to use `.ts` files after this.  comment annotations have limitations, like not allowing `as SomeInterface`.  i find it to be cumbersome the way it is in `.js` files...

vasco-santos commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,146 @@
+        try {
+          const evaluateJson = await googleCloudResponse.json()
+          const stringifiedJSON = JSON.stringify(evaluateJson)
+          await env.CID_VERIFIER_RESULTS.put(resultKey, stringifiedJSON)
```
```suggestion
          await env.CID_VERIFIER_RESULTS.put(resultKey, stringifiedJSON, { metadata: evaluateJson })
```

vasco-santos commented on packages/edge-gateway/wrangler.toml:
```diff
@@ -29,11 +29,11 @@ routes = [
 [env.production.vars]
 IPFS_GATEWAYS = "[\"https://ipfs.io\", \"https://cf.dag.haus\"]"
 GATEWAY_HOSTNAME = 'ipfs.dag.haus'
+CID_VERIFIER_URL = 'http://cid-verifier.localhost:8787'
```
```suggestion
CID_VERIFIER_URL = 'https://cid-verifier.dag.haus'
```

vasco-santos commented on packages/edge-gateway/wrangler.toml:
```diff
@@ -71,11 +77,11 @@ routes = [
 [env.staging.vars]
 IPFS_GATEWAYS = "[\"https://ipfs.io\", \"https://cf.dag.haus\", \"https://pinata.dag.haus\"]"
 GATEWAY_HOSTNAME = 'ipfs-staging.dag.haus'
+CID_VERIFIER_URL = 'http://cid-verifier.localhost:8787'
```
```suggestion
CID_VERIFIER_URL = 'https://cid-verifier-staging.dag.haus
```

jsdevel commented on packages/cid-verifier/src/verification.js:
```diff
@@ -0,0 +1,146 @@
+        env.log.log(`GOOGLE CLOUD UNABLE TO VERIFY URL "${url}"https://cloud.google.com/web-risk/docs/reference/rest/v1eap1/TopLevel/evaluateUri#ConfidenceLevel status code "${googleCloudResponse?.status}"`, 'error')
+      }
+
+      if (!resultsStored) {
+        // delete PENDING so we can try again
+        await env.CID_VERIFIER_RESULTS.delete(resultKey)
+      }
```
if for whatever reason there was an error and we didn't write to either the result key or DENYLIST i figured we'd want to try again right?

jsdevel commented on packages/edge-gateway/test/index.spec.js:
```diff
@@ -101,3 +101,10 @@ test('Gets content with other base encodings', async (t) => {
   t.is(await response.text(), 'Hello dot.storage! 😎')
 })
+
+test('Sends HTML files to cid-verifier', async (t) => {
```
@vasco-santos i'm having trouble getting this to work.  can you take a look?

jsdevel reviewed (commented):
Hoping you can get this sample.html test to work.

vasco-santos reviewed (approved):
LGTM 🎉
